### PR TITLE
Make p2tr payments and taproot.ts more PSBT friendly

### PR DESCRIPTION
**Pull request by brandonblack**

### CODE REVIEW

File: ts_src/taproot.ts [hunk 1]
```diff
@@ -186,6 +193,42 @@ interface WeightedTapScript {
...
+  const { value, done } = leaves.next();
+  assert(!done, 'insufficient leaves to reconstruct tap tree');
+  const [index, leaf] = value;
+  const tree: Taptree = {
+    root: hashTapLeaf(leaf.script, leaf.leafVersion),
+    paths: [],
+  };
+  tree.paths[index] = [];
+  for (let depth = leaf.depth; depth > targetDepth; depth--) {
+    const sibling = recurseTaptree(leaves, depth);
```
Comment (sangaman) [inline ts_src/taproot.ts:209]: If we're on the last leaf, won't this throw? Because we'll go back into this function, do `leaves.next()`, and then `done` should be true, unless I'm mistaken which is quite possible.

In other words, what happens if there is no sibling here?

File: ts_src/taproot.ts [hunk 2]
```diff
@@ -186,6 +193,42 @@ interface WeightedTapScript {
...
+    tree.paths.forEach(path => path.push(sibling.root));
+    sibling.paths.forEach(path => path.push(tree.root));
+    tree.root = hashTapBranch(tree.root, sibling.root);
+    tree.paths = Object.assign(tree.paths, sibling.paths);
```
Comment (sangaman) [inline ts_src/taproot.ts:213]: I actually didn't know you could combine arrays with `Object.assign` (tho it makes sense since arrays are just objects). FWIW, I think `Object.assign()` modifies the `tree.paths` variable in place so the assignment is unnecessary I believe. Alternatively you could do `tree.paths = tree.paths.concat(sibling.paths);` But I think these are all equivalent, just thinking out loud.

**Review by sangaman — APPROVED** [review]
LGTM, just a few questions. The `recurseTapTree` logic is a little hard to follow but I don't have any great suggestions to improve it, it might just be the nature of depth-first trees.

File: ts_src/payments/p2tr.ts [hunk 3]
```diff
@@ -86,8 +87,16 @@ export function p2tr(a: Payment, opts?: PaymentOpts): Payment {
...
-  const _taptree = lazy.value(() => {
+
+  const network = a.network || BITCOIN_NETWORK;
+
+  const o: Payment = { network };
+
+  const _taprootPaths = lazy.value(() => {
     if (!a.redeems) return;
+    if (o.tapTree) {
+      return taproot.getDepthFirstTaptree(o.tapTree);
```
[the remark this answers is not in the document]
Comment (brandonblack) [reply ts_src/payments/p2tr.ts:98]: The `o.tapTree` property will be undefined (and print a deprecation warning) if the input was weight-based.

File: ts_src/payments/p2tr.ts [hunk 4]
```diff
@@ -108,15 +117,16 @@ export function p2tr(a: Payment, opts?: PaymentOpts): Payment {
...
+  lazy.prop(o, 'internalPubkey', () => {
     if (a.pubkey) {
       // single pubkey
       return a.pubkey;
     } else if (a.pubkeys && a.pubkeys.length === 1) {
       return a.pubkeys[0];
     } else if (a.pubkeys && a.pubkeys.length > 1) {
       // multiple pubkeys
-      return taproot.aggregateMuSigPubkeys(ecc, a.pubkeys);
+      return Buffer.from(taproot.aggregateMuSigPubkeys(ecc, a.pubkeys));
```
[the remark this answers is not in the document]
Comment (brandonblack) [reply ts_src/payments/p2tr.ts:129]: I'm not fundamentally against it, but that bit of code is going to be replaced when we do musig signing later this year, so don't want to touch it now.

File: ts_src/taproot.ts (hunk 1 above)
Comment (brandonblack) [reply ts_src/taproot.ts:209]: If there is no sibling, it will throw. If we enter here because `depth` is still less than `targetDepth` and there are no siblings, the tree was invalid. There must be sufficient siblings to reach the `targetDepth`

File: ts_src/taproot.ts (hunk 2 above)
Comment (brandonblack) [reply ts_src/taproot.ts:213]: Good call on skipping the unnecessary assignment - at first I had `Object.assign([], ...)` so then the assignment was necessary.

I can't do concat, because they are potentially disjoint sparse arrays depending on the recursion path... I think.

Comment (brandonblack) [reply ts_src/taproot.ts]: Removed unnecessary assignment, added comment.
